File: leetcode_problems/python/lc322_coin_change.py
class SolutionGreedyWrong:
    def coinChange(
        self,
        coins: list[int],
        amount: int,
    ) -> int:
        coins.sort(reverse=True)
        count = 0
        index = 0
        while amount > 0:
            if index >= len(coins):
                return -1
            if coins[index] > amount:
                index += 1
                continue
            amount -= coins[index]
            count += 1
        return count


class Solution:
    def coinChange(
        self,
        coins: list[int],
        amount: int,
    ) -> int:
        n = amount + 1
        dp = [10**5] * n
        dp[0] = 0
        for coin in coins:
            for i in range(coin, n):
                dp[i] = min(dp[i], dp[i - coin] + 1)
        return dp[amount] if dp[amount] != 10**5 else -1


if __name__ == "__main__":
    coins = [5, 2, 1]
    amount = 11

    print(Solution().coinChange(coins, amount))

File: leetcode_problems/cpp/lc322_coin_change.cpp
#include <vector>
#include <cmath>
using std::vector;

class Solution
{
public:
    int coinChange(vector<int> &coins, int amount)
    {
        int n = amount + 1;
        vector<int> dp(n, std::pow(10, 5));
        dp[0] = 0;
        for (int coin : coins)
        {
            for (int i = coin; i < n; i++)
            {
                dp[i] = std::min(dp[i], dp[i - coin] + 1);
            }
        }
        return dp[amount] == std::pow(10, 5) ? -1 : dp[amount];
    }
};